File: accbot-android/app/src/main/java/com/accbot/dca/domain/usecase/CalculatePortfolioUseCase.kt
package com.accbot.dca.domain.usecase

import com.accbot.dca.data.local.TransactionEntity
import com.accbot.dca.domain.model.Exchange
import java.math.BigDecimal
import java.math.RoundingMode
import java.time.YearMonth
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import javax.inject.Inject

/**
 * Data class representing crypto holdings.
 */
data class CryptoHolding(
    val crypto: String,
    val totalAmount: BigDecimal,
    val totalInvested: BigDecimal,
    val averagePrice: BigDecimal,
    val transactionCount: Int
)

/**
 * Data class representing holdings on a specific exchange.
 */
data class ExchangeHolding(
    val exchange: Exchange,
    val holdings: List<CryptoHolding>,
    val totalInvested: BigDecimal
)

/**
 * Data class representing monthly performance.
 */
data class MonthlyPerformance(
    val month: String,           // "Jan 2024"
    val yearMonth: String,       // "2024-01"
    val totalInvested: BigDecimal,
    val totalCrypto: BigDecimal,
    val transactionCount: Int,
    val averagePrice: BigDecimal
)

/**
 * Data class containing all portfolio calculations.
 */
data class PortfolioSummary(
    val cryptoHoldings: List<CryptoHolding>,
    val exchangeHoldings: List<ExchangeHolding>,
    val monthlyPerformance: List<MonthlyPerformance>,
    val totalInvested: BigDecimal,
    val totalBtc: BigDecimal,
    val totalTransactions: Int,
    val averageMonthlyInvestment: BigDecimal
)

/**
 * Use case for portfolio calculations.
 * Extracts business logic from ViewModel for better testability and separation of concerns.
 */
class CalculatePortfolioUseCase @Inject constructor() {

    companion object {
        private val monthDisplayFormatter = DateTimeFormatter.ofPattern("MMM yyyy")
    }

    /**
     * Calculate complete portfolio summary from transactions.
     */
    fun calculatePortfolioSummary(transactions: List<TransactionEntity>): PortfolioSummary {
        val cryptoHoldings = calculateCryptoHoldings(transactions)
        val exchangeHoldings = calculateExchangeHoldings(transactions)
        val monthlyPerformance = calculateMonthlyPerformance(transactions)

        val totalInvested = transactions.sumOf { it.fiatAmount }
        val totalBtc = transactions.filter { it.crypto == "BTC" }.sumOf { it.cryptoAmount }
        val totalTransactions = transactions.size

        val averageMonthly = if (monthlyPerformance.isNotEmpty()) {
            totalInvested.divide(BigDecimal(monthlyPerformance.size), 2, RoundingMode.HALF_UP)
        } else {
            BigDecimal.ZERO
        }

        return PortfolioSummary(
            cryptoHoldings = cryptoHoldings,
            exchangeHoldings = exchangeHoldings,
            monthlyPerformance = monthlyPerformance,
            totalInvested = totalInvested,
            totalBtc = totalBtc,
            totalTransactions = totalTransactions,
            averageMonthlyInvestment = averageMonthly
        )
    }

    /**
     * Calculate holdings grouped by cryptocurrency.
     * Time complexity: O(n) where n = number of transactions.
     */
    fun calculateCryptoHoldings(transactions: List<TransactionEntity>): List<CryptoHolding> {
        return transactions
            .groupBy { it.crypto }
            .map { (crypto, txs) ->
                val totalAmount = txs.sumOf { it.cryptoAmount }
                val totalInvested = txs.sumOf { it.fiatAmount }
                val averagePrice = calculateAveragePrice(totalInvested, totalAmount)

                CryptoHolding(
                    crypto = crypto,
                    totalAmount = totalAmount,
                    totalInvested = totalInvested,
                    averagePrice = averagePrice,
                    transactionCount = txs.size
                )
            }
            .sortedByDescending { it.totalInvested }
    }

    /**
     * Calculate holdings grouped by exchange.
     * Time complexity: O(n) where n = number of transactions.
     */
    fun calculateExchangeHoldings(transactions: List<TransactionEntity>): List<ExchangeHolding> {
        return transactions
            .groupBy { it.exchange }
            .map { (exchange, txs) ->
                val holdings = txs
                    .groupBy { it.crypto }
                    .map { (crypto, cryptoTxs) ->
                        val totalAmount = cryptoTxs.sumOf { it.cryptoAmount }
                        val totalInvested = cryptoTxs.sumOf { it.fiatAmount }
                        val averagePrice = calculateAveragePrice(totalInvested, totalAmount)

                        CryptoHolding(
                            crypto = crypto,
                            totalAmount = totalAmount,
                            totalInvested = totalInvested,
                            averagePrice = averagePrice,
                            transactionCount = cryptoTxs.size
                        )
                    }

                ExchangeHolding(
                    exchange = exchange,
                    holdings = holdings.sortedByDescending { it.totalInvested },
                    totalInvested = txs.sumOf { it.fiatAmount }
                )
            }
            .sortedByDescending { it.totalInvested }
    }

    /**
     * Calculate monthly performance metrics.
     * Time complexity: O(n) where n = number of transactions.
     */
    fun calculateMonthlyPerformance(transactions: List<TransactionEntity>): List<MonthlyPerformance> {
        return transactions
            .groupBy { tx ->
                val localDate = tx.executedAt.atZone(ZoneId.systemDefault()).toLocalDate()
                YearMonth.from(localDate)
            }
            .map { (yearMonth, txs) ->
                val totalInvested = txs.sumOf { it.fiatAmount }
                val totalCrypto = txs.filter { it.crypto == "BTC" }.sumOf { it.cryptoAmount }
                val averagePrice = calculateAveragePrice(totalInvested, totalCrypto)

                MonthlyPerformance(
                    month = yearMonth.format(monthDisplayFormatter),
                    yearMonth = yearMonth.toString(),
                    totalInvested = totalInvested,
                    totalCrypto = totalCrypto,
                    transactionCount = txs.size,
                    averagePrice = averagePrice
                )
            }
            .sortedByDescending { it.yearMonth }
    }

    /**
     * Calculate average price with safe division.
     */
    private fun calculateAveragePrice(totalFiat: BigDecimal, totalCrypto: BigDecimal): BigDecimal {
        return if (totalCrypto > BigDecimal.ZERO) {
            totalFiat.divide(totalCrypto, 2, RoundingMode.HALF_UP)
        } else {
            BigDecimal.ZERO
        }
    }
}

File: accbot-ios/AccBot/Domain/UseCase/CalculatePortfolioUseCase.swift
import Foundation

struct CryptoHolding {
    let crypto: String
    let totalAmount: Decimal
    let totalInvested: Decimal
    let averagePrice: Decimal
    let transactionCount: Int
}

struct ExchangeHolding {
    let exchange: Exchange
    let holdings: [CryptoHolding]
    let totalInvested: Decimal
}

struct MonthlyPerformance {
    let month: String        // "Jan 2024"
    let yearMonth: String    // "2024-01"
    let totalInvested: Decimal
    let totalCrypto: Decimal
    let transactionCount: Int
    let averagePrice: Decimal
}

struct PortfolioSummary {
    let cryptoHoldings: [CryptoHolding]
    let exchangeHoldings: [ExchangeHolding]
    let monthlyPerformance: [MonthlyPerformance]
    let totalInvested: Decimal
    let totalBtc: Decimal
    let totalTransactions: Int
    let averageMonthlyInvestment: Decimal
}

/// Calculates portfolio holdings, exchange breakdowns, and monthly performance from transactions.
final class CalculatePortfolioUseCase {

    func calculatePortfolioSummary(transactions: [Transaction]) -> PortfolioSummary {
        let cryptoHoldings = calculateCryptoHoldings(transactions: transactions)
        let exchangeHoldings = calculateExchangeHoldings(transactions: transactions)
        let monthlyPerformance = calculateMonthlyPerformance(transactions: transactions)

        let totalInvested = transactions.reduce(Decimal.zero) { $0 + $1.fiatAmount }
        let totalBtc = transactions.filter { $0.crypto == "BTC" }.reduce(Decimal.zero) { $0 + $1.cryptoAmount }
        let totalTransactions = transactions.count

        let averageMonthly: Decimal = if !monthlyPerformance.isEmpty {
            safeDiv(totalInvested, by: Decimal(monthlyPerformance.count), scale: 2)
        } else {
            0
        }

        return PortfolioSummary(
            cryptoHoldings: cryptoHoldings,
            exchangeHoldings: exchangeHoldings,
            monthlyPerformance: monthlyPerformance,
            totalInvested: totalInvested,
            totalBtc: totalBtc,
            totalTransactions: totalTransactions,
            averageMonthlyInvestment: averageMonthly
        )
    }

    func calculateCryptoHoldings(transactions: [Transaction]) -> [CryptoHolding] {
        Dictionary(grouping: transactions, by: \.crypto)
            .map { (crypto, txs) in
                let totalAmount = txs.reduce(Decimal.zero) { $0 + $1.cryptoAmount }
                let totalInvested = txs.reduce(Decimal.zero) { $0 + $1.fiatAmount }
                return CryptoHolding(
                    crypto: crypto,
                    totalAmount: totalAmount,
                    totalInvested: totalInvested,
                    averagePrice: safeDiv(totalInvested, by: totalAmount, scale: 2),
                    transactionCount: txs.count
                )
            }
            .sorted { $0.totalInvested > $1.totalInvested }
    }

    func calculateExchangeHoldings(transactions: [Transaction]) -> [ExchangeHolding] {
        Dictionary(grouping: transactions, by: \.exchange)
            .map { (exchange, txs) in
                let holdings = Dictionary(grouping: txs, by: \.crypto)
                    .map { (crypto, cryptoTxs) in
                        let totalAmount = cryptoTxs.reduce(Decimal.zero) { $0 + $1.cryptoAmount }
                        let totalInvested = cryptoTxs.reduce(Decimal.zero) { $0 + $1.fiatAmount }
                        return CryptoHolding(
                            crypto: crypto,
                            totalAmount: totalAmount,
                            totalInvested: totalInvested,
                            averagePrice: safeDiv(totalInvested, by: totalAmount, scale: 2),
                            transactionCount: cryptoTxs.count
                        )
                    }
                    .sorted { $0.totalInvested > $1.totalInvested }

                return ExchangeHolding(
                    exchange: exchange,
                    holdings: holdings,
                    totalInvested: txs.reduce(Decimal.zero) { $0 + $1.fiatAmount }
                )
            }
            .sorted { $0.totalInvested > $1.totalInvested }
    }

    func calculateMonthlyPerformance(transactions: [Transaction]) -> [MonthlyPerformance] {
        let calendar = Calendar.current
        let monthFormatter = DateFormatter()
        monthFormatter.dateFormat = "MMM yyyy"
        let yearMonthFormatter = DateFormatter()
        yearMonthFormatter.dateFormat = "yyyy-MM"

        return Dictionary(grouping: transactions) { tx in
            let components = calendar.dateComponents([.year, .month], from: tx.executedAt)
            return "\(components.year!)-\(String(format: "%02d", components.month!))"
        }
        .map { (ym, txs) in
            let totalInvested = txs.reduce(Decimal.zero) { $0 + $1.fiatAmount }
            let totalCrypto = txs.reduce(Decimal.zero) { $0 + $1.cryptoAmount }

            let sampleDate = txs.first!.executedAt
            let displayMonth = monthFormatter.string(from: sampleDate)

            return MonthlyPerformance(
                month: displayMonth,
                yearMonth: ym,
                totalInvested: totalInvested,
                totalCrypto: totalCrypto,
                transactionCount: txs.count,
                averagePrice: safeDiv(totalInvested, by: totalCrypto, scale: 2)
            )
        }
        .sorted { $0.yearMonth > $1.yearMonth }
    }

    private func safeDiv(_ a: Decimal, by b: Decimal, scale: Int) -> Decimal {
        guard b > 0 else { return 0 }
        let handler = NSDecimalNumberHandler(
            roundingMode: .plain, scale: Int16(scale),
            raiseOnExactness: false, raiseOnOverflow: false,
            raiseOnUnderflow: false, raiseOnDivideByZero: false
        )
        return NSDecimalNumber(decimal: a)
            .dividing(by: NSDecimalNumber(decimal: b), withBehavior: handler)
            .decimalValue
    }
}
